controlplane/kubeadm/internal/cluster.go

```diff
@@ -0,0 +1,383 @@
+	return &metav1.LabelSelector{
+		MatchLabels: m.ControlPlaneLabelsForCluster(clusterName),
+	}
+}
+
+// cluster are operations on target clusters.
+type cluster struct {
+	client ctrlclient.Client
+	// restConfig is required for the proxy.
+	restConfig            *rest.Config
+	etcdCACert, etcdCAkey []byte
+}
+
+// generateEtcdTLSClientBundle builds an etcd client TLS bundle from the Etcd CA for this cluster.
+func (c *cluster) generateEtcdTLSClientBundle() (*tls.Config, error) {
+	clientCert, err := generateClientCert(c.etcdCACert, c.etcdCAkey)
+	if err != nil {
+		return nil, err
+	}
+
+	caPool := x509.NewCertPool()
+	caPool.AppendCertsFromPEM(c.etcdCACert)
+
+	return &tls.Config{
+		RootCAs:      caPool,
+		Certificates: []tls.Certificate{clientCert},
+	}, nil
+}
+
+// etcdIsHealthyResponse is a map of node provider IDs to the etcd health check that failed or nil if no checks failed.
+type etcdIsHealthyResponse map[string]error
+
+// etcdIsHealthy runs checks for every etcd member in the cluster to satisfy our definition of healthy.
+// This is a best effort check and nodes can become unhealthy after the check is complete. It is not a guarantee.
+// It's used a signal for if we should allow a target cluster to scale up, scale down or upgrade.
+// It returns a list of nodes checked so the layer above can confirm it has the correct number of nodes expected.
+func (c *cluster) etcdIsHealthy(ctx context.Context) (etcdIsHealthyResponse, error) {
+	var knownClusterID uint64
+	var knownMemberIDSet etcdutil.UInt64Set
+
+	controlPlaneNodes := &corev1.NodeList{}
+	controlPlaneNodeLabels := map[string]string{
+		"node-role.kubernetes.io/master": "",
+	}
+
+	if err := c.client.List(ctx, controlPlaneNodes, client.MatchingLabels(controlPlaneNodeLabels)); err != nil {
+		return nil, err
+	}
+
+	tlsConfig, err := c.generateEtcdTLSClientBundle()
+	if err != nil {
+		return nil, err
+	}
+
+	response := map[string]error{}
+	for _, node := range controlPlaneNodes.Items {
+		name := node.Name
+		response[name] = nil
+		if node.Spec.ProviderID == "" {
+			response[name] = errors.New("empty provider ID")
+			continue
+		}
+
+		// Create the etcd client for the etcd Pod scheduled on the Node
+		etcdClient, err := c.getEtcdClientForNode(name, tlsConfig)
+		if err != nil {
+			response[name] = errors.New("failed to create etcd client")
+			continue
+		}
+
+		// List etcd members. This checks that the member is healthy, because the request goes through consensus.
+		members, err := etcdClient.Members(ctx)
+		if err != nil {
+			response[name] = errors.New("failed to list etcd members using etcd client")
+			continue
+		}
+		member := etcdutil.MemberForName(members, name)
+
+		// Check that the member reports no alarms.
+		if len(member.Alarms) > 0 {
+			response[name] = errors.Errorf("etcd member reports alarms: %v", member.Alarms)
+			continue
+		}
+
+		// Check that the member belongs to the same cluster as all other members.
+		clusterID := member.ClusterID
```

Comment:
the `knownClusterID` var needs to exist outside of the `for` loop. Since it's a constant across all etcd members, we want to make sure they all match.

Reply:
Makes sense!